src/Generator/ConvertGenerator.php

```diff
@@ -91,10 +90,14 @@ public static function getSupportedTypes() {
 	/** @var EpubGenerator */
 	private $epubGenerator;
 
-	public function __construct( int $timeout, FileCache $fileCache, EpubGenerator $epubGenerator ) {
+	/** @var ?Semaphore */
+	private $semaphore;
+
+	public function __construct( int $timeout, FileCache $fileCache, EpubGenerator $epubGenerator, ?Semaphore $semaphore = null ) {
 		$this->timeout = $timeout;
 		$this->fileCache = $fileCache;
 		$this->epubGenerator = $epubGenerator;
+		$this->semaphore = $semaphore;
```

Comment:
I think there might be a problem with injecting a Semaphore here in that it'll be instantiated for every request (including those that don't use ConvertGenerator, e.g. just opening the main export form). Then `sem_get()` will be called every time, and won't that mean that it'll run out of lock acquisitions? An alternative might be to inject a SemaphoreFactory or similar, or move `sem_get()` into `tryLock()`?

Reply:
`sem_get` only creates the semaphore if it does not exists yet and returns its resource descriptor but does not lock it. Locking is done with `sem_acquire`. I have updated the MR to lazily call `sem_get` only on `tryLock()` to avoid running it if the semaphore is not used. Thank you!

Reply:
Ah makes sense, thanks! And looks good now.